### PR TITLE
Remove const enums from API

### LINKED ISSUE #5219: Const enums in the TS Compiler API can make depending on typescript difficult
At present, if I write a library which depends on typescript, I implicitly lock my consumers into using a typescript package version which has a `SyntaxKind` which matches the numbers which existed when I _compiled_ my library, and failures resulting from a mismatch in runtime and compile-time `SyntaxKind`s only occur as unexpected runtime behavior (like a tslint `no-null` rule triggering on `if` statements). This behavior change may only even be noticed when the package is consumed and not when it's tested (since, frequently, you test with the same TS you build with).

I would suppose that fixing this could require either considering changes to our API's const enums as semver breaking changes to make it easier on npm, or we should stop using const enums in our API, so the name-based references are retained in consumers' compiled output. Alternatively, we _could_ continue to use const enums internally, but in our built `.d.ts` we report them all as normal enums (and we compile with `preserveConstEnums`, as we already do), so our consumers retain the references in _their_ built code.

### PR DESCRIPTION
**Pull request by weswigham**
Adds a compiler flag to emit const enums as normal enums in the declaration emit (requiring that they be preserved in the js emit), then enables it on the compiler.

Enabling the flag on the compiler itself required an LKG update - There's another LKG update (without this flag) pending at #5293.

Fixes #5219.

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
:+1:

File: src/compiler/diagnosticMessages.json [hunk 1]
```diff
@@ -2125,6 +2125,10 @@
         "category": "Message",
         "code": 6020
     },
+    "Indicates that const enum delcarations are to be emitted as enum declarations. Requires --preserveConstEnums.": {
```
Comment (sandersn) [inline src/compiler/diagnosticMessages.json]: Typo: declarations

File: src/compiler/commandLineParser.ts [hunk 2]
```diff
@@ -148,7 +148,12 @@ namespace ts {
         {
             name: "preserveConstEnums",
             type: "boolean",
-            description: Diagnostics.Do_not_erase_const_enum_declarations_in_generated_code
+            description: Diagnostics.Do_not_erase_const_enum_declarations_in_generated_code,
+        },
+        {
+            name: "deconstConstEnums",
+            type: "boolean",
+            description: Diagnostics.Indicates_that_const_enum_delcarations_are_to_be_emitted_as_enum_declarations_Requires_preserveConstEnums,
```
Comment (sandersn) [inline src/compiler/commandLineParser.ts]: Typo: declarations

**Comment by sandersn** [comment]
I can't come up with any better option names given the current command line parser design. If we ever have a chance to redo the command line parser design, I'd like to see something like

```
--constEnum=
  emit -- emit code for const enums but do not use the emitted code anywhere
  preserve -- emit and preserve enum references for consumers of this compilation's d.ts.
  inline -- inline const enums everywhere [default]
```

Then `--constEnum=emit` could be the equivalent of today's `--preserveConstEnum`.

**Comment by sandersn** [comment]
:+1:

**Comment by RyanCavanaugh** [comment]
Very hesitant on having a flag like this vs just a post-build script that uses the AST to strip out the `const` keyword. Let's discuss at the design meeting.

**Comment by mhegazy** [comment]
i agree with @RyanCavanaugh i do not think we need another flag. If you really want this behavior just do not use const enums.

**Comment by weswigham** [comment]
@mhegazy should we stop using const enums in the compiler?

**Comment by mhegazy** [comment]
> @mhegazy should we stop using const enums in the compiler?

i did not say that :). we have two scenarios that are a bit contradicting, one in the compiler, we want to avoid the perf hit from the extra property lookup, and in the public API, we would like consumers to use the enum. 

For our use cases, I think a post build step to clean up the .d.ts is better than adding yet another enum variant in the language.

**Comment by weswigham** [comment]
Alright, though by doing so we acknowledge it's a shortcoming in our compiler's declaration emit which is meant to be worked around with an external tool - I don't doubt that other libraries may have this issue and will have to have similar postbuild declaration processing.

**Comment by mhegazy** [comment]
I am not sure where the shortcoming is to be frank. but there are already too many enum concepts out there that necessary.

**Comment by weswigham** [comment]
> I am not sure where the shortcoming is to be frank.
It's in how we do const inlining and then expose that const inlining was done via dts. Const enums work well in a big-bang ts build, where all of your ts deps get built at the same time against one another. When versions start to drift between compile time and build time, you lose the ability to safely inline const enums, since you can't guarantee that the library A you build your TS against is the library A you're running against at runtime (because the JS world is one of late binding).

For performance reasons, people enjoy const inlining (as we do), and so we have const enums, but for library purposes, you can't safely use const enums in your APIs (without monitoring them for even the smallest changes, so you know to increment a semver change indicative of the break). Part of me doubts we'll be the only TS library that will ever have this issue, and should others have it we'll not be the only ones who will want to solve it.

But I suppose it can always be revisited later, if/when other people bring it up, so it doesn't really matter for now.

**Pull request closed without merging**